confgenerator/confgenerator.go

```diff
@@ -97,6 +99,144 @@ func (m *Metrics) generateOtelPipelines() (map[string]otel.Pipeline, error) {
 	return out, nil
 }
 
+func (uc *UnifiedConfig) GenerateFluentBitConfigsModular(logsDir string, stateDir string, hostInfo *host.InfoStat) (string, string, error) {
+	userAgent, _ := getUserAgent("Google-Cloud-Ops-Agent-Logging", hostInfo)
+	components, err := uc.Logging.generateFluentbitComponents(userAgent, hostInfo)
+	if err != nil {
+		return "", "", err
+	}
+
+	c := fluentbit.ModularConfig{
+		Variables: map[string]string{
+			"buffers_dir": path.Join(stateDir, "buffers"),
+			"logs_dir":    logsDir,
+		},
+		Components: components,
+	}
+	return c.Generate()
+}
+
+type fbSource struct {
+	tag        string
+	components []fluentbit.Component
+}
+
+func (l *Logging) generateFluentbitComponents(userAgent string, hostInfo *host.InfoStat) ([]fluentbit.Component, error) {
+	var out []fluentbit.Component
+	out = append(out, fluentbit.Service{}.Component())
+
+	if l != nil && l.Service != nil {
+		var sources []fbSource
+		var logNames []string
+		for pID, p := range l.Service.Pipelines {
+			for _, rID := range p.ReceiverIDs {
+				receiver, ok := l.Receivers[rID]
+				if !ok {
+					return nil, fmt.Errorf("receiver %q not found", rID)
+				}
+				tag := fmt.Sprintf("%s.%s", pID, rID)
+				components := receiver.Components(tag)
+				for i, pID := range p.ProcessorIDs {
+					processor, ok := l.Processors[pID]
+					if !ok {
+						return nil, fmt.Errorf("processor %q not found", pID)
+					}
+					components = append(components, processor.Components(tag, i)...)
```

Comment:
Just curious, why wouldn't you use the processor id in the parser name? It might be easier than counting to track where the parser config came from. Unless the same processor id can appear more than once in a pipeline, but I suspect that would be an invalid config…

Reply:
I believe that's a technically valid config.

Reply:
A little too meta, but fine.